net/pfSense-pkg-WireGuard/files/usr/local/pkg/wireguard/wg_guiconfig.inc

```diff
@@ -123,50 +132,75 @@ function wg_handshake_status_icon($latest_handshake = 0, $fa_icon = "fa-handshak
 
 	}
 
+	wg_htmlspecialchars($fa_icon, $ret_class, $ret_title);
+
 	$ret_html = "<i class=\"fa {$fa_icon} {$ret_class}\" style=\"vertical-align: middle;\" title=\"{$ret_title}\"></i>";
 
 	return $ret_html;
 
 }
 
 // Formats a given time difference in a human-friendly way
-function wg_human_time_diff($start_time, $end_time = null, $from_epoch = false) {
-	
-	// If end time is not supplied, use the current time instead
-	$end_time = is_null($end_time) ? time() : $end_time;
+function wg_human_time_diff($start_time_stamp, $end_time_stamp = null, $from_epoch = false) {
+
+	if (is_null($end_time_stamp)) {
+
+		$end_time = new DateTimeImmutable();
+
+	} elseif (is_numericint($end_time_stamp)) {
+
+		$end_time = new DateTime();
+
+		$end_time->setTimestamp($end_time_stamp);
+
+	}
+
+	$start_time = new DateTime();
+
+	if ($start_time_stamp > 0 || $from_epoch) {
+
+		$start_time->setTimestamp($start_time_stamp);
+
+	} else {
 
-	// If start time is zero, proceed unless we are actually interested in calculating time from epoch, probably not...
-	$start_time = ($start_time > 0) ? $start_time : ($from_epoch ? $start_time : $end_time);
+		return gettext('never');
 
-	$diff_time = abs($end_time - $start_time);
+	}
+
+	$diff_time = $end_time->diff($start_time);
+
+	$diff_time_seconds = abs($end_time->getTimestamp() - $start_time->getTimestamp());
 
 	$tokens = array (
-		31536000 => 'year',
-		2592000 => 'month',
-		604800 => 'week',
-		86400 => 'day',
-		3600 => 'hour',
-		60 => 'minute',
-		1 => 'second'
+		31536000 => gettext('year'),
+		2592000 => gettext('month'),
+		604800 => gettext('week'),
+		86400 => gettext('day'),
+		3600 => gettext('hour'),
+		60 => gettext('minute'),
+		1 => gettext('second')
 	);
 
 	foreach ($tokens as $unit => $text) {
 
-		if ($diff_time < $unit) {
-			
+		if ($diff_time_seconds < $unit) {
+
 			continue;
 			
 		}
 
-		$num_of_units = floor($diff_time / $unit);
+		$num_of_units = floor($diff_time_seconds / $unit);
 
 		$plural = ($num_of_units > 1) ? 's' : '';
+
+		$tense = ($diff_time->invert) ? 'ago' : 'from now';
 		
-		return "{$num_of_units} {$text}{$plural}";
+		return "{$num_of_units} {$text}{$plural} {$tense}";
```

Comment:
Not terribly fond of this part. We have a similar function already, `convert_seconds_to_dhms()` https://github.com/pfsense/pfsense/blob/master/src/etc/inc/pfsense-utils.inc#L1575 seems like that would be a better fit, or you could leverage the PHP `DateInterval::format` method at https://www.php.net/manual/en/dateinterval.format.php which can output closer to what you are doing manually without reinventing the wheel.

Reply:
What I'm doing here is more closely aligning to the handshake format displayed in various WireGuard clients. From what I've seen, this is likely as good as it can get in order to get this particular output style. I'm using DateTime to derive the number of seconds and then some basic string manipulation to get the desired final output. For what it's worth, this is not too far off from how `wg(8)` generates the latest handshake time as well.

Reply:
https://git.zx2c4.com/wireguard-tools/tree/src/show.c#n129

Reply:
I get why they do it that way in C, but in PHP it makes less sense, as the DateTime library has methods for handling exactly that type of output and formatting. There are better examples on the php.net page I linked above which do similar things without hardcoding the various time values.

Reply:
Since contributed the original code here I will comment, I took a look at  those examples. There all use very similar code to get exactly what we want to do. This is another example https://www.php.net/manual/en/dateinterval.format.php#102271 that uses almost identical logic

Reply:
I guess the question here is really, do we want latest handshake times to be displayed in the "WireGuard" way (i.e. how WireGuard clients display it) or in a more  "pfSense" way (i.e. how these types of times are displayed in other parts of the UI). I'm obviously more inclined to the former. I'm not immediately seeing how the `DateInterval::format` method can be used as a replacement for implementing the former as this method doesn't return the total seconds and will always require maths to convert and add-up the various denominations.

Reply:
@jim-p I'll look at the example you linked more closely in a few hours. A cursory glance looks promising, because what we are effectively doing is only showing the 'largest' denomination and that example is pretty easy to adjust to produce that. Thanks.

Reply:
Alright, spent some quality time on this tonight. I think we are square on this now.